### ISSUE TITLE
"invalid column number in pdo_sqlsrv_stmt_get_col_data" and PHP crash when using variable length tables in query

### ISSUE THREAD

**Issue by thirdwheel** [microsoft/msphpsql#1217]
+## PHP Driver version or file name
 +Driver version 5.8.0
 +## SQL Server version
 +SQL Server 2016
 +## Client operating system
 +Windows Server 2016 Standard
 +## PHP version
 +7.3.10
 +## Microsoft ODBC Driver version
 +2017.175.01.01
 +## Table schema
SQL:
``` sql
CREATE OR ALTER PROCEDURE [dbo].[pdobugtest] (@level int) AS
BEGIN
    IF @level = 1
        SELECT 1 Col1
              ,1 Col2
              ,1 Col3
              ,1 Col4
              ,1 Col5
              ,1 Col6
              ,1 Col7
              ,1 Col8
              ,1 Col9
              ,1 Col10
              ,1 Col11
              ,1 Col12
              ,1 Col13
              ,1 Col14
              ,1 Col15
              ,1 Col16
              ,1 Col17
              ,1 Col18
              ,1 Col19
              ,1 Col20

    IF @Level = 2
        SELECT 1 Col1
              ,1 Col2
              ,1 Col3
              ,1 Col4
              ,1 Col5
              ,1 Col6
              ,1 Col7
              ,1 Col8
              ,1 Col9
              ,1 Col10
              ,1 Col11
              ,1 Col12
              ,1 Col13
              ,1 Col14

END
GO
```
 +## Problem description

It appears that if you have a stored procedure that returns different numbers of columns based on parameters, the run with the higher number of columns causes PHP to crash.  If I recreate the statement each time I need to use it, this error goes away and the script succeeds.

 +## Expected behavior and actual behavior

Expected behaviour - both runs succeed and return results
Actual behaviour - first run succeeds, second causes PHP to crash

 +## Repro code or steps to reproduce
 
1. Create stored procedure

2. Change the DSN to match your environment and run PHP code below:

``` php
<?php

$db=new PDO('sqlsrv:Server=SERVER\INSTANCE;Database=dbname;MultipleActiveResultSets=0;', NULL, NULL);

$x=$db->prepare(
    "exec pdobugtest ?"
);
$x->execute([2]);
$x->fetch();
$x->closeCursor();

$x->execute([1]);
$x->fetch();
$x->closeCursor();
```

**Comment by thirdwheel** [comment]
Addendum - this issue still persists if I update to 5.9.0 preview 1

**Comment by yitam** [comment]
Thanks @thirdwheel for the report. We will get back to you.

**Comment by yitam** [comment]
Hi @thirdwheel 

While it's a good idea to use [prepared statements](https://stackoverflow.com/questions/3385177/best-practices-with-preparedstatements-when-to-and-when-not-to), your approach is actually problematic, mainly because different number of columns is returned depending on your input. 

FYI, in pdo_stmt.c [line 506](https://github.com/php/php-src/blob/php-7.3.25/ext/pdo/pdo_stmt.c#L506), after the first execution, the stmt object keeps the resultset's columns and their metadata, so the second time you execute again they would not be re-populated. Not only that, since the number of columns is now 20, when the stmt object is going through its destruction phase, it thinks there are 20 column objects to be destroyed. I tested this in a debugging mode it actually crashed at the 15th column, as you can easily imagine.

In short, you're probably better off with direct queries because the entire column metadata after the first execution is basically useless to be kept for next execution.

**Comment by thirdwheel** [comment]
Thanks, I had a suspicion that was what was happening based on my reading of that part of the code. Unfortunately using direct queries is out of the question due to utilising user inputs.  I'll stick to my current approach - recreating the prepared statement on each run to clear the stored metadata.

**Comment by yitam** [comment]
Thanks @thirdwheel for your prompt response. Prepared statements are resilient against SQL injection, so you have valid concerns. Just a suggestion, based on your scenario, instead of using a prepared statement to invoke a stored procedure that takes a parameter, perhaps put the logic in your application. That is, you can have more than one prepared statement and invoke the correct one depending on the user inputs, In that case, there's no need to recreate the prepared statements every single time. 

Please feel free to reopen this if necessary.

**Comment by thirdwheel** [comment]
Thanks for the suggestion, however the high variability in the data coming back means an exponential number of prepared statements would need to be created, so recreating the single prepared statement seems like it'd be more efficient.

Either that, or have logic to check the number of columns (not hard since the number of columns is deterministic) and process them in descending order. I found if I reversed the order in the example no errors occurred.  I just tried this and it works.

Posting here in case anyone else stumbles on this issue and this is a solution that works for them.
